R/FilterStateExpr.R

```diff
@@ -0,0 +1,350 @@
+    #' library(shinyjs)
+    #'
+    #' ui <- fluidPage(
+    #'   useShinyjs(),
+    #'   include_css_files(pattern = "filter-panel"),
+    #'   include_js_files(pattern = "count-bar-labels"),
+    #'   column(4, div(
+    #'     h4("ChoicesFilterState"),
+    #'     isolate(filter_state$ui("fs"))
+    #'   )),
+    #'   column(8, div(
+    #'     h4("Condition (i.e. call)"), # display the subsetting call generated by this FilterState
+    #'     textOutput("condition_choices"), br(),
+    #'     h4("Unformatted state"), # display raw filter state
+    #'     textOutput("unformatted_choices"), br(),
+    #'     h4("Formatted state"), # display human readable filter state
+    #'     textOutput("formatted_choices"), br()
+    #'   ))
+    #' )
+    #'
+    #' server <- function(input, output, session) {
+    #'   filter_state$server("fs")
+    #'   output$condition_choices <- renderPrint(filter_state$get_call())
+    #'   output$formatted_choices <- renderText(filter_state$format())
+    #'   output$unformatted_choices <- renderPrint(filter_state$get_state())
+    #' }
+    #'
+    #' if (interactive()) {
+    #'   shinyApp(ui, server)
+    #' }
+    #' }
+    #'
+    #' @return `FilterStateExpr`
+    initialize = function(id, title, dataname, expr, disabled = FALSE, ...) {
+      checkmate::assert_string(id)
+      checkmate::assert_string(title)
+      checkmate::assert_string(dataname)
+      assert_logical_expr(expr)
+      checkmate::assert_flag(disabled)
+
+      private$id <- id
+      private$title <- title
+      private$dataname <- dataname
+      private$expr <- expr
+      private$disabled <- reactiveVal(disabled)
+      private$extras <- list(...)
+      invisible(self)
+
+    },
+
+    #' @description
+    #' Returns a formatted string representing this `FilterState`.
+    #'
+    #' @param indent (`numeric(1)`)
+    #'   number of spaces before after each new line character of the formatted string;
+    #'   defaults to 0
+    #' @param wrap_width (`numeric(1)`)
+    #'   number of characters to wrap lines at in the printed output;
+    #'   allowed range is 30 to 120; defaults to 76
+    #'
+    #' @return `character(1)` the formatted string
+    #'
+    format = function(indent = 2L, wrap_width = 76L) {
+      checkmate::assert_number(indent, finite = TRUE, lower = 0L)
+      checkmate::assert_number(wrap_width, finite = TRUE, lower = 30L, upper = 120L)
+
+      # List all selected values separated by commas.
+      paste(
+        strwrap(
+          sprintf("Filter expr: %s", deparse1(private$expr)),
+          width = wrap_width,
+          indent = indent
+        ),
+        collapse = "\n")
+    },
+
+    #' @description
+    #' Returns filtering state.
+    #'
+    #' @return A `teal_slice` object.
+    #'
+    get_state = function() {
+      states <- append(
+        list(
+          id = private$id,
+          title = private$title,
+          dataname = private$dataname,
+          expr = quote(private$expr),
```

Comment:
Since `private$expr` stores `expr` and `expr` must be an expression, is `quote` necessary?

Reply:
It must be "quotes" as `expr` is stored as language, and `do.call` would try evaluate this if not quoted.